pkg/kv/kvserver/raft_transport.go

```diff
@@ -610,12 +939,99 @@ func (t *RaftTransport) startProcessNewQueue(
 		return false
 	}
 	return true
 }
 
+// startDroppingFlowTokensForDisconnectedNodes kicks of an asynchronous worker
+// that periodically scans for nodes we're no longer connected to, and if there
+// are any pending flow tokens bound for that node, it simply drops them. This
+// "connected nodes" is a client-side view of the world. On the server-side when
+// gRPC streams disconnect, we release all held tokens. So simply dropping these
+// pending dispatches on the client side does not cause token leaks, and exists
+// to prevent an unbounded accumulation of memory.
+func (t *RaftTransport) startDroppingFlowTokensForDisconnectedNodes(ctx context.Context) error {
+	return t.stopper.RunAsyncTask(
+		ctx,
+		"kvserver.RaftTransport: drop flow tokens for disconnected nodes",
+		func(ctx context.Context) {
+			settingChangeCh := make(chan struct{}, 1)
+			kvadmission.FlowTokenDropInterval.SetOnChange(
+				&t.st.SV, func(ctx context.Context) {
+					select {
+					case settingChangeCh <- struct{}{}:
+					default:
+					}
+				})
+
+			timer := timeutil.NewTimer()
+			defer timer.Stop()
+
+			for {
+				interval := kvadmission.FlowTokenDropInterval.Get(&t.st.SV)
+				if interval > 0 {
+					timer.Reset(interval)
+				} else {
+					// Disable the mechanism.
+					timer.Stop()
+					timer = timeutil.NewTimer()
+				}
+				select {
+				case <-timer.C:
+					timer.Read = true
+					t.dropFlowTokensForDisconnectedNodes()
+					continue
+
+				case <-settingChangeCh:
+					// Loop around to use the updated timer.
+					continue
+
+				case <-t.stopper.ShouldQuiesce():
+					return
+				}
+			}
+		},
+	)
+}
+
+// TestingDropFlowTokensForDisconnectedNodes exports
+// dropFlowTokensForDisconnectedNodes for testing purposes.
+func (t *RaftTransport) TestingDropFlowTokensForDisconnectedNodes() {
+	t.dropFlowTokensForDisconnectedNodes()
+}
+
+// TestingPrintFlowControlConnectionTracker renders the state of the underlying
+// connection tracker.
+func (t *RaftTransport) TestingPrintFlowControlConnectionTracker() string {
+	t.kvflowControl.mu.Lock()
+	defer t.kvflowControl.mu.Unlock()
+	return t.kvflowControl.mu.connectionTracker.testingPrint()
+}
+
+func (t *RaftTransport) dropFlowTokensForDisconnectedNodes() {
+	t.kvflowControl.mu.Lock()
+	defer t.kvflowControl.mu.Unlock()
+	for _, nodeID := range t.kvflowControl.dispatchReader.PendingDispatch() {
+		if t.kvflowControl.mu.connectionTracker.isNodeConnected(nodeID) {
+			// If there's already a queue active, there's nothing to do. We rely
+			// on timely piggybacking of flow tokens on existing raft transport
+			// messages. It's only when all queues are idle/culled that we use
+			// this worker to drop any held tokens. The expectation is that on
+			// the server-side of the stream, we'll have returned all tokens
+			// when streams disconnect.
+			continue
+		}
+		// Drop any held tokens for that node.
+		pendingDispatches := t.kvflowControl.dispatchReader.PendingDispatchFor(nodeID)
```

Comment:
`t.kvflowControl.dispatchReader` is not protected by mu, we only have access to the reader portion of the interface, the writes are happening elsewhere. And there's no race -- we're dealing with the nodeID we know is not connected in the loop very iteration we're in.

> We loop around and not drop the newly added dispatches

Two things:
- When looping around, we won't see the nodeID we already processed, again.
- Any newly added dispatches for that nodeID that need to be dropped (because we're still not connected to it) will be dropped `kvadmission.flow_token.drop_interval `(30s) later through the higher level `startDroppingFlowTokensForDisconnectedNodes` loop.